From: Matthias Runge <mrunge@redhat.com>
Date: Tue, 26 Nov 2013 08:47:39 +0000 (+0100)
Subject: add support for Django-1.6
X-Git-Url: https://review.openstack.org/gitweb?p=openstack%2Fhorizon.git;a=commitdiff_plain;h=8d376ad83be0237105d109b9209ff157a64a758f

add support for Django-1.6

Sessions now store JSON instead of pickled objects.

Partially-implements blueprint django-1point6

Change-Id: I332ba18005284774a53fb3cb8c6e51bca1480ded
---

diff --git a/horizon/middleware.py b/horizon/middleware.py
index 61fcdbb..7dabe13 100644
--- a/horizon/middleware.py
+++ b/horizon/middleware.py
@@ -21,9 +21,9 @@
 Middleware provided and used by Horizon.
 """
 
-import datetime
 import json
 import logging
+import time
 
 from django.conf import settings  # noqa
 from django.contrib.auth import REDIRECT_FIELD_NAME  # noqa
@@ -61,11 +61,12 @@ class HorizonMiddleware(object):
             timeout = 1800
 
         last_activity = request.session.get('last_activity', None)
-        timestamp = datetime.datetime.now()
+        timestamp = int(time.time())
         request.horizon = {'dashboard': None,
                            'panel': None,
                            'async_messages': []}
-        if last_activity and (timestamp - last_activity).seconds > timeout:
+        if isinstance(last_activity, int) and \
+                 (timestamp - last_activity) > timeout:
             request.session.pop('last_activity')
             response = HttpResponseRedirect(
                 '%s?next=%s' % (settings.LOGOUT_URL, request.path))
diff --git a/horizon/test/tests/middleware.py b/horizon/test/tests/middleware.py
index d480fa1..6f3051b 100644
--- a/horizon/test/tests/middleware.py
+++ b/horizon/test/tests/middleware.py
@@ -15,7 +15,7 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
-import datetime
+import time
 
 from django.conf import settings  # noqa
 
@@ -44,7 +44,7 @@ class MiddlewareTests(test.TestCase):
         except AttributeError:
             timeout = 1800
         request.session['last_activity'] =\
-            datetime.datetime.now() - datetime.timedelta(seconds=timeout + 10)
+            int(time.time()) - (timeout + 10)
         mw = middleware.HorizonMiddleware()
         resp = mw.process_request(request)
         self.assertEqual(resp.status_code, 302)
diff --git a/openstack_dashboard/settings.py b/openstack_dashboard/settings.py
index 1216099..2944164 100644
--- a/openstack_dashboard/settings.py
+++ b/openstack_dashboard/settings.py
@@ -171,6 +171,9 @@ SESSION_COOKIE_HTTPONLY = True
 SESSION_EXPIRE_AT_BROWSER_CLOSE = True
 SESSION_COOKIE_SECURE = False
 SESSION_TIMEOUT = 1800
+# when doing upgrades, it may be wise to stick to PickleSerializer
+# TODO(mrunge): remove after Icehouse
+SESSION_SERIALIZER = 'django.contrib.sessions.serializers.PickleSerializer'
 
 gettext_noop = lambda s: s
 LANGUAGES = (
